File: apps/mlx-qwen3-omni/src/mlx_qwen3_omni/audio/mel.py
"""
Mel spectrogram for Qwen3-Omni Audio Encoder (AuT).

Computes 128-bin log-mel spectrogram from raw 16 kHz mono audio.
Matches Whisper-style mel extraction: Hann window, STFT (hop=160, win=400),
mel filterbank (0–8 kHz), log-mel. Reference: apps/rust-perf/src/candle_mel.rs
"""

from __future__ import annotations

import math
from typing import Union

import numpy as np

# Default STFT hop length (samples).
HOP_LENGTH = 160
# Default window length (samples).
N_FFT = 400
# Number of mel bins (Qwen3-Omni audio_config.num_mel_bins).
N_MELS = 128
# Sample rate (Hz).
SAMPLE_RATE = 16_000
# Mel scale: high frequency bound (Hz).
FMAX = 8000.0
# Mel scale: low frequency bound (Hz).
FMIN = 0.0


def _hz_to_mel(hz: float) -> float:
    return 2595.0 * math.log10(1.0 + hz / 700.0)


def _mel_to_hz(mel: float) -> float:
    return 700.0 * (10.0 ** (mel / 2595.0) - 1.0)


def _build_mel_filterbank(
    n_mels: int,
    n_freq: int,
    sample_rate: int,
    fmin: float,
    fmax: float,
) -> np.ndarray:
    """Build mel filterbank matrix (n_mels, n_freq). Slaney-style triangular filters."""
    sr = float(sample_rate)
    mel_low = _hz_to_mel(fmin)
    mel_high = _hz_to_mel(fmax)
    mel_points = np.linspace(mel_low, mel_high, n_mels + 2)
    hz_points = np.array([_mel_to_hz(float(m)) for m in mel_points])
    bin_points = (n_freq - 1) * hz_points / (sr / 2.0)

    mel_fb = np.zeros((n_mels, n_freq), dtype=np.float32)
    for i in range(n_mels):
        left = bin_points[i]
        center = bin_points[i + 1]
        right = bin_points[i + 2]
        for j in range(n_freq):
            jf = float(j)
            if jf < left or jf > right:
                v = 0.0
            elif jf < center:
                v = (jf - left) / max(center - left, 1e-6)
            else:
                v = (right - jf) / max(right - center, 1e-6)
            mel_fb[i, j] = v
    return mel_fb


class MelSpectrogram:
    """
    Pre-allocated mel spectrogram extractor.
    Output shape: (1, n_mels, time_frames) to match Rust candle_mel.
    """

    def __init__(
        self,
        n_fft: int = N_FFT,
        hop_length: int = HOP_LENGTH,
        n_mels: int = N_MELS,
        sample_rate: int = SAMPLE_RATE,
        fmin: float = FMIN,
        fmax: float = FMAX,
    ) -> None:
        self.n_fft = n_fft
        self.hop_length = hop_length
        self.n_mels = n_mels
        self.sample_rate = sample_rate
        self.fmin = fmin
        self.fmax = fmax
        n_freq = n_fft // 2 + 1
        self._window = np.hanning(n_fft).astype(np.float32)
        self._mel_fb = _build_mel_filterbank(
            n_mels, n_freq, sample_rate, fmin, fmax
        )

    def compute(self, samples: Union[np.ndarray, list[float]]) -> np.ndarray:
        """
        Compute log-mel spectrogram from raw audio samples.
        Returns array of shape (1, n_mels, time_frames).
        """
        if isinstance(samples, list):
            samples = np.array(samples, dtype=np.float32)
        else:
            samples = np.asarray(samples, dtype=np.float32)
        if samples.ndim != 1:
            raise ValueError("samples must be 1D (mono)")
        n_freq = self.n_fft // 2 + 1
        n_frames = 1 + max(0, (samples.size - self.n_fft) // self.hop_length)
        if n_frames == 0:
            return np.zeros((1, self.n_mels, 1), dtype=np.float32)

        # STFT: frame-by-frame with Hann window
        mel_out = np.zeros((n_frames, self.n_mels), dtype=np.float32)
        for t in range(n_frames):
            start = t * self.hop_length
            frame = np.zeros(self.n_fft, dtype=np.float32)
            end = min(start + self.n_fft, samples.size)
            frame[: end - start] = samples[start:end] * self._window[: end - start]
            # FFT (real) -> power spectrum
            fft_out = np.fft.rfft(frame)
            power = (fft_out.real ** 2 + fft_out.imag ** 2).astype(np.float32)
            # Mel filterbank
            mel_out[t, :] = self._mel_fb @ power

        # Log-mel (log1p to avoid log(0))
        mel_out = np.log1p(mel_out)
        # (n_frames, n_mels) -> (1, n_mels, n_frames)
        return mel_out.T[np.newaxis, :, :]


def compute_mel(
    samples: Union[np.ndarray, list[float]],
    n_fft: int = N_FFT,
    hop_length: int = HOP_LENGTH,
    n_mels: int = N_MELS,
    sample_rate: int = SAMPLE_RATE,
    fmin: float = FMIN,
    fmax: float = FMAX,
) -> np.ndarray:
    """
    One-shot log-mel spectrogram from raw audio.
    Returns (1, n_mels, time_frames).
    """
    mel = MelSpectrogram(
        n_fft=n_fft,
        hop_length=hop_length,
        n_mels=n_mels,
        sample_rate=sample_rate,
        fmin=fmin,
        fmax=fmax,
    )
    return mel.compute(samples)

File: apps/rust-mlx-omni/src/mel.rs
//! Mel spectrogram: raw audio → mel filterbank features.
//!
//! Port of `apps/mlx-qwen3-omni/src/mlx_qwen3_omni/audio/mel.py`.
//! Reference: `apps/rust-perf/src/candle_mel.rs`.

use mlx_rs::{ops, Array};

/// Compute mel spectrogram from raw audio samples.
///
/// - `audio`: (samples,) f32, 16kHz mono.
/// - Returns: (1, num_mel_bins, time_frames) suitable for the audio encoder.
pub fn mel_spectrogram(
    audio: &Array,
    n_fft: usize,
    hop_length: usize,
    num_mel_bins: usize,
    sample_rate: usize,
) -> Result<Array, mlx_rs::error::Exception> {
    // 1. STFT via Hann window
    let n_fft_i = n_fft as i32;
    let hop_i = hop_length as i32;

    // Hann window
    let hann = hann_window(n_fft)?;

    // Pad audio to center frames
    let pad_len = n_fft / 2;
    let pad_width: [(i32, i32); 1] = [(pad_len as i32, pad_len as i32)];
    let audio = ops::pad(audio, &pad_width, None, None)?;

    // Frame the audio: extract overlapping windows
    let num_samples = audio.shape()[0];
    let num_frames = (num_samples - n_fft_i) / hop_i + 1;

    // Use as_strided to create frames efficiently
    let frames = create_frames(&audio, n_fft, hop_length, num_frames as usize)?;

    // Apply window and compute FFT magnitude
    let windowed = frames.multiply(&hann)?;
    let spectrum = mlx_rs::fft::rfft(&windowed, Some(n_fft_i), -1)?;

    // Power spectrum: |S|^2
    let real = spectrum.real()?;
    let imag = spectrum.imag()?;
    let power = real.multiply(&real)?.add(&imag.multiply(&imag)?)?;

    // 2. Mel filterbank
    let mel_basis = mel_filterbank(sample_rate, n_fft, num_mel_bins)?;

    // (num_frames, n_fft/2+1) @ (n_fft/2+1, num_mel_bins) → (num_frames, num_mel_bins)
    let mel_basis_t = mel_basis.transpose_axes(&[1, 0])?;
    let mel_spec = power.matmul(&mel_basis_t)?;

    // Log mel: log(max(mel, 1e-10))
    let floor = Array::from(1e-10_f32);
    let mel_spec = ops::maximum(&mel_spec, &floor)?;
    let log_mel = mel_spec.log()?;

    // Reshape to (1, num_mel_bins, time_frames)
    let log_mel = log_mel.transpose_axes(&[1, 0])?; // (mel, time)
    Ok(log_mel.expand_dims(0)?) // (1, mel, time)
}

/// Hann window of length n.
fn hann_window(n: usize) -> Result<Array, mlx_rs::error::Exception> {
    let n_f = n as f32;
    let pi2 = 2.0 * std::f32::consts::PI;
    let values: Vec<f32> = (0..n)
        .map(|i| 0.5 * (1.0 - (pi2 * i as f32 / n_f).cos()))
        .collect();
    Ok(Array::from_slice(&values, &[n as i32]))
}

/// Extract overlapping frames from audio signal.
fn create_frames(
    audio: &Array,
    frame_len: usize,
    hop_len: usize,
    num_frames: usize,
) -> Result<Array, mlx_rs::error::Exception> {
    // Build index array for gather
    let mut indices: Vec<i32> = Vec::with_capacity(num_frames * frame_len);
    for f in 0..num_frames {
        let start = f * hop_len;
        for i in 0..frame_len {
            indices.push((start + i) as i32);
        }
    }
    let idx = Array::from_slice(&indices, &[num_frames as i32, frame_len as i32]);
    // Flatten audio, gather, reshape
    let flat = audio.reshape(&[-1])?;
    let idx_flat = idx.reshape(&[-1])?;
    Ok(flat
        .take_axis(&idx_flat, 0)?
        .reshape(&[num_frames as i32, frame_len as i32])?)
}

/// Create mel filterbank matrix (num_mel_bins, n_fft/2+1).
fn mel_filterbank(
    sample_rate: usize,
    n_fft: usize,
    num_mel_bins: usize,
) -> Result<Array, mlx_rs::error::Exception> {
    let n_freqs = n_fft / 2 + 1;
    let sr = sample_rate as f64;

    // Mel scale: mel = 2595 * log10(1 + f/700)
    let f_min = 0.0_f64;
    let f_max = sr / 2.0;
    let mel_min = 2595.0 * (1.0 + f_min / 700.0).log10();
    let mel_max = 2595.0 * (1.0 + f_max / 700.0).log10();

    // Mel points
    let mel_points: Vec<f64> = (0..num_mel_bins + 2)
        .map(|i| mel_min + (mel_max - mel_min) * i as f64 / (num_mel_bins + 1) as f64)
        .collect();

    // Convert back to Hz
    let hz_points: Vec<f64> = mel_points
        .iter()
        .map(|&m| 700.0 * (10.0_f64.powf(m / 2595.0) - 1.0))
        .collect();

    // Convert to FFT bin indices
    let bin_points: Vec<f64> = hz_points
        .iter()
        .map(|&f| f * n_fft as f64 / sr)
        .collect();

    // Build filterbank
    let mut fb = vec![0.0_f32; num_mel_bins * n_freqs];
    for m in 0..num_mel_bins {
        let left = bin_points[m];
        let center = bin_points[m + 1];
        let right = bin_points[m + 2];

        for k in 0..n_freqs {
            let kf = k as f64;
            if kf >= left && kf <= center {
                fb[m * n_freqs + k] = ((kf - left) / (center - left)) as f32;
            } else if kf > center && kf <= right {
                fb[m * n_freqs + k] = ((right - kf) / (right - center)) as f32;
            }
        }
    }

    Ok(Array::from_slice(
        &fb,
        &[num_mel_bins as i32, n_freqs as i32],
    ))
}
